### PR TITLE
Bugfix/sim 1664/pointing

### PR DESCRIPTION
**Pull request by danielsf**
This pull request fixes an ambiguity in what we mean by 'RA' and 'Dec'.

Going forward, all routines that accept RA and Dec as arguments will expect coordinates in the International Celestial Reference System (ICRS), meaning that precession, nutation, aberration, and parallax (all the lovely effects that we have to worry about because the Earth is moving) will not be included.  The methods appGeoFromICRS and observedFromAppGeo will apply these effects before transforming from ICRS to focal plane or pixel coordinates.  Furthermore, ObservationMetaData will store the pointing in ICRS.

This slows down a bunch of our catalog functionality, since all of the coordinates need to be transformed from ICRS to observed coordinates before camera position is calculated, but that is the only way to compute position on the camera accurately.

There are supporting pull requests in

sims_catalogs_generation
sims_catalogs_measures
sims_coordUtils
sims_catUtils
sims_GalSimInterface
sims_maf

### CODE REVIEW

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    the site of the telescope and the MJD of the observation
+
+    @param [out] altitude in degrees
+
+    @param [out] azimuth in degrees
+
+    @param [out] parallactic angle in degrees
+    """
+
+    # TODO time This should be passing around Universal Time
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Is this note still relevant?

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+    @param [in] az is the azimuth in degrees.  Cant be a numpy array or a single value.
+
+    @param [in] obs is an ObservationMetaData characterizing
+    the site of the telescope and the MJD of the observation
+
+    @param [out] RA in degrees (in the International Celestial Reference System)
+
+    @param [out] Dec in degrees (in the International Celestial Reference System)
+
+    Note: This method is only accurate to within 0.01 arcsec near azimuth = 0 or pi
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Why is this?

Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Actually, since this was already here, if this is discussed in a previous pull request, never mind.

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+    WARNING: As of 13 April 2015, this method does not agree with OpSim on
+    the relationship between rotSkyPos and rotTelPos.  This is due to a
+    discrepancy between the time that OpSim uses as the MJD when calculating
+    rotTelPos and the time that OpSim reports as being the actual expmjd
+    of the exposure (rotTelPos is calculated at the beginning of the exposure;
+    expmjd is reckoned at the middle of the exposure).
+    """
+    altRad, azRad, paRad = _altAzPaFromRaDec(raRad, decRad, obs)
+
+    #20 March 2015
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Is this still the case?

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+    #I do not know where this expression comes from; we should validate it against
+    #the definitions of rotTelPos and rotSkyPos
+    return (rotSkyRad + paRad)%(2.*np.pi)
+
+
+def calcObsDefaults(raRad, decRad, altRad, azRad, rotTelRad, mjd, band, longRad, latRad):
+    """
+    Fromats input data into a dict of metadata that PhoSim expect.
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Typo Formats

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+
+
+def makeObservationMetadata(metaData):
+    return OrderedDict([(k,(metaData[k], np.asarray(metaData[k]).dtype))
+                         for k in metaData])
+
+
+def makeObsParamsAzAltTel(azRad, altRad, mjd, band, rotTelRad=0., longRad=-1.2320792, latRad=-0.517781017, **kwargs):
+    '''
+    Calculate a minimal set of observing parameters give the ra, dec, and time of the observation.
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Shouldn't this and the first line in the docstring for the next method say "given the alt, az, and time of the observation." rather than ra, dec, time?

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+__all__ = ["_pupilCoordsFromRaDec", "pupilCoordsFromRaDec",
+           "_raDecFromPupilCoords", "raDecFromPupilCoords"]
+
+
+def pupilCoordsFromRaDec(ra_in, dec_in, obs_metadata=None, epoch=None):
+    """
+    Take an input RA and dec from the sky and convert it to coordinates
+    on the focal plane.
+
+    This uses PAL's gnonomonic projection routine which assumes that the focal
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: typo gnomonic

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    return _pupilCoordsFromRaDec(np.radians(ra_in), np.radians(dec_in),
+                                 obs_metadata=obs_metadata, epoch=epoch)
+
+
+def _pupilCoordsFromRaDec(ra_in, dec_in, obs_metadata=None, epoch=None):
+    """
+    Take an input RA and dec from the sky and convert it to coordinates
+    on the focal plane.
+
+    This uses PAL's gnonomonic projection routine which assumes that the focal
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: typo gnomonic

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    transforms those RA, Dec into observed geocentric coordinates, applying
+    the effects of precession, nutation, aberration, parallax and refraction.
+    This is done, because the gnomonic projection ought to be applied to what
+    observers actually see, rather than the idealized, above-the-atmosphere
+    coordinates represented by the ICRS.
+
+    @param [in] ra_in is a numpy array of RAs in degrees
+    (in the International Celestial Reference System)
+
+    @param [in] dec_in in degrees
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: Also mention that it is a numpy array like ra_in?

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+    # The extra negative sign on x_out comes from the following:
+    # The Gnomonic projection as calculated by palpy is such that,
+    # if north is in the +y direction, then west is in the -x direction,
+    # which is the opposite of the behavior we want (I do not know how to
+    # express this analytically; I have just confirmed it numerically)
+    x *= -1.0
+
+    #rotate the result by rotskypos (rotskypos being "the angle of the sky relative to
+    #camera cooridnates" according to phoSim documentation) to account for
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: typo coordinates

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    """
+    @param [in] xPupil -- pupil coordinates in radians
+
+    @param [in] yPupil -- pupil coordinates in radians
+
+    @param [in] obs_metadata -- an instantiation of ObservationMetaData characterizing
+    the state of the telescope
+
+    @param [in] epoch -- julian epoch of the mean equinox used for the coordinate
+    transforations (in years)
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: typo transformations

File: python/lsst/sims/utils/FocalPlaneUtils.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+    x_g = xPupil*np.cos(theta) - yPupil*np.sin(theta)
+    y_g = xPupil*np.sin(theta) + yPupil*np.cos(theta)
+
+    x_g *= -1.0
+
+    # x_g and y_g are now the x and y coordinates
+    # can now use the PALPY method palDtp2s to convert to RA, Dec.
+
+    raObs, decObs = palpy.dtp2sVector(x_g, y_g, ra_pointing, dec_pointing)
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: Should there be a backup loop here like in the reverse transformation in case one of the array ra/dec values doesn't work?

File: python/lsst/sims/utils/WcsUtils.py [hunk 12]
```diff
@@ -1,103 +1,231 @@
-import numpy
+import numpy as np
```
Comment (jbkalmbach) [inline python/lsst/sims/utils/WcsUtils.py]: Why did you change it here but leave it as numpy in other files?

File: tests/testAstrometry.py [hunk 13]
```diff
@@ -311,98 +440,203 @@ def test_applyPrecession(self):
...
+                                    numpy.power(xyz_control[2]-z_list_icrs,2))
+
+            dex = numpy.argmax(distance)
+            msg = 'pm %e %e vr %e px %e; time %e; err %e arcsec; corr %e' % \
+            (arcsecFromRadians(pm_ra[dex]), arcsecFromRadians(pm_dec[dex]),
+             v_rad[dex], arcsecFromRadians(px[dex]), pmt, arcsecFromRadians(distance[dex]),
+             arcsecFromRadians(correction[dex]))
+
+            self.assertLess((distance/correction).max(), 0.01, msg=msg)
+            # demand that the two methods agree on the stars' new positions to within one part in 100
```
Comment (jbkalmbach) [inline tests/testAstrometry.py]: Is there a reason you picked this level of accuracy? Why is it good enough?

File: tests/testAstrometry.py [hunk 14]
```diff
@@ -311,98 +440,203 @@ def test_applyPrecession(self):
...
+            ra_test, dec_test = _appGeoFromICRS(numpy.array([ra_icrs]), numpy.array([dec_icrs]),
+                                                mjd=obs.mjd,
+                                                pm_ra=numpy.array([pm_ra]),
+                                                pm_dec=numpy.array([pm_dec]),
+                                                v_rad=numpy.array([v_rad]),
+                                                parallax=numpy.array([px]),
+                                                epoch=2000.0)
+
+            distance = arcsecFromRadians(haversine(ra_app, dec_app, ra_test[0], dec_test[0]))
+            self.assertLess(distance, 0.1)
```
Comment (jbkalmbach) [inline tests/testAstrometry.py]: Same question here as in other test. Where do definitions of acceptable error come from?

File: tests/testCompoundCoordinateTransformations.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+        latitude_list.append(np.radians(11.0+45.0/60.0))
+        mjd_list.append(2457364.958333-2400000.5) # 8 December 2015 11:00 UTC
+        alt_list.append(np.radians(41.1))
+        az_list.append(np.radians(134.7))
+        ra_app_list.append(16.0*hours + 59.0*minutes + 16.665*seconds)
+        dec_app_list.append(np.radians(-22.0 - 42.0/60.0 - 2.94/3600.0))
+
+        longitude_list.append(np.radians(-22.0-33.0/60.0))
+        latitude_list.append(np.radians(11.0+45.0/60.0))
+        mjd_list.append(2457368.958333-2400000.5) # 12 Decemer 2015 11:00 UTC
```
Comment (jbkalmbach) [inline tests/testCompoundCoordinateTransformations.py]: Typo December

File: tests/testCompoundCoordinateTransformations.py [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+                                                            latitude=latitude),
+                                            mjd=utils.ModifiedJulianDate(UTC=mjd))
+
+
+            ra_icrs, dec_icrs = utils._raDecFromAltAz(alt, az, obs)
+            ra_test, dec_test = utils._appGeoFromICRS(np.array([ra_icrs]), np.array([dec_icrs]),
+                                                      mjd=obs.mjd)
+
+            distance = np.degrees(utils.haversine(ra_app, dec_app, ra_test[0], dec_test[0]))
+            self.assertLess(distance, 0.1) # since that is all the precision we have in alt az
```
Comment (jbkalmbach) [inline tests/testCompoundCoordinateTransformations.py]: But the note in the method docstring says we should have 0.01 not 0.1.

File: tests/testPupilCoordinates.py [hunk 17]
```diff
@@ -162,14 +173,15 @@ def testRaDecFromPupil(self):
...
         numpy.random.seed(42)
         ra = (numpy.random.random_sample(nSamples)*0.1-0.2) + numpy.radians(raCenter)
         dec = (numpy.random.random_sample(nSamples)*0.1-0.2) + numpy.radians(decCenter)
         xp, yp = _pupilCoordsFromRaDec(ra, dec, obs_metadata=obs, epoch=2000.0)
         raTest, decTest = _raDecFromPupilCoords(xp, yp, obs_metadata=obs, epoch=2000.0)
-        numpy.testing.assert_array_almost_equal(raTest, ra, decimal=10)
-        numpy.testing.assert_array_almost_equal(decTest, dec, decimal=10)
+        distance = arcsecFromRadians(haversine(ra, dec, raTest, decTest))
+        self.assertLess(distance.max(), 1.0) # because of the imprecision in _icrsFromObserved,
+                                             # the best we can hope for is sub-arcsecond accuracy
```
Comment (jbkalmbach) [inline tests/testPupilCoordinates.py]: Just want to make sure this is good enough.

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py (hunk 1 above)
Comment (danielsf) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Good point.  No, it is not relevant.

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py (hunk 2 above)
Comment (danielsf) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Honestly, I'm not sure.  I think I just observed that behavior when I was writing unit tests for it.  If I had to guess, I would say it comes from ```haRad = np.where(np.sin(azRad)>=0.0, -1.0*haRad0, haRad0)```, but it seems like, if that were going to cause problems, it would do so at the greater than 0.01 arcsecond level.  I'll open an issue to investigate the accuracy of this method.

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py (hunk 3 above)
Comment (danielsf) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: Technically, yes it is true in that I don't know where the relationship comes from, but we have checked it against OpSim (as mentioned in the docstring).  I will go ahead and remove the comment.

File: python/lsst/sims/utils/CompoundCoordinateTransformations.py (hunk 5 above)
Comment (danielsf) [inline python/lsst/sims/utils/CompoundCoordinateTransformations.py]: true

File: python/lsst/sims/utils/FocalPlaneUtils.py (hunk 11 above)
Comment (danielsf) [inline python/lsst/sims/utils/FocalPlaneUtils.py]: I don't think so.  The backup loop in pupilCoordinatesFromRaDec is in case you accidentally ask for a point that is so far from your boresite that the transformation from spherical coordinates to coordinates on the focal plane breaks, at which point palpy will raise an exception (so: I am pointing at RA=25, Dec=-11 and I ask for the coordinates of a star at RA=125, Dec=50).  I feel like, if you are going the other way, you are unlikely to accidentally specify focal plane coordinates that are so far from your boresite that transformation back to spherical coordinates breaks.

File: python/lsst/sims/utils/WcsUtils.py (hunk 12 above)
Comment (danielsf) [inline python/lsst/sims/utils/WcsUtils.py]: Because I didn't want to flood this PR with a bunch of small changes that meant nothing.  I adopted the philosophy "I will change whatever file I touch".  Once we have converged on this pull request, I will go through and change everything else.  That way, no one has to look at a cluttered file diff.

File: tests/testAstrometry.py (hunk 13 above)
Comment (danielsf) [inline tests/testAstrometry.py]: I chose it because that is the level of accuracy at which the test passes.

Given that PALPY is not modular in the way that it handles astrometric corrections (i.e. the method that transforms from mean to apparent geocentric position does not call the method which applies proper motion, even though the transformation from mean to apparent geocentric position include proper motion effects), I don't think there is a way to get these to agree exactly.  I have chosen not to sweat a difference of one part in 100.

I acknowledge that is not the most satisfying answer.

File: tests/testAstrometry.py (hunk 14 above)
Comment (danielsf) [inline tests/testAstrometry.py]: The definitions come from my gut.  There is no documentation anywhere guaranteeing that the data I got from NOAO is actually presenting the same quantities that PALPY is caclulating.  My threshold for these unit tests is "do I believe the results", and, at the 0.1 arcsecond level, I have decided that I do.

Again: not very satisfying.

File: tests/testCompoundCoordinateTransformations.py (hunk 16 above)
Comment (danielsf) [inline tests/testCompoundCoordinateTransformations.py]: Sorry.  That was confusing.

The 0.1 is the accuracy in Alt, Az that we have in the data I downloaded from NOAO.  I will clarify the comment.

Comment (danielsf) [inline tests/testCompoundCoordinateTransformations.py]: excuse me: the USNO

File: tests/testPupilCoordinates.py (hunk 17 above)
Comment (danielsf) [inline tests/testPupilCoordinates.py]: numpy random was generating points too near the Sun.  I changed it so that the test points are all more than 45 degrees from the sun.  Now the accuracy is within 0.005 arcseconds.

**Comment by jbkalmbach** [comment]
Ok, this looks good.
